### PR TITLE
DX: check deprecations exactly

### PR DESCRIPTION
**Pull request by kubawerlos**
Ref. https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7578#discussion_r1434021534

### CODE REVIEW

File: tests/Test/AbstractIntegrationTestCase.php [hunk 1]
```diff
@@ -143,6 +143,8 @@ public function testIntegration(IntegrationCase $case): void
     {
         foreach ($case->getSettings()['deprecations'] as $deprecation) {
             $this->expectDeprecation($deprecation);
+            $this->expectDeprecation($deprecation);
+            $this->expectDeprecation($deprecation);
```
Comment (Wirone) [inline tests/Test/AbstractIntegrationTestCase.php:147]: Does it make sense to expect exactly the same deprecation message multiple times? I would rather want to ensure that every expected deprecation message happens at least once. What is the gain from checking the exact amount of messages?

File: tests/Console/ConfigurationResolverTest.php [hunk 2]
```diff
@@ -1207,13 +1207,15 @@ public static function provideDeprecatedFixerConfiguredCases(): iterable
      */
     public function testDeprecatedRuleSetConfigured(string $ruleSet, array $successors): void
     {
-        $this->expectDeprecation(sprintf(
+        $this->expectDeprecation($message = sprintf(
```
Comment (keradus) [inline tests/Console/ConfigurationResolverTest.php:1210]: please do not assign within function call, but before.

File: tests/Test/AbstractIntegrationTestCase.php (hunk 1 above)
Comment (keradus) [reply tests/Test/AbstractIntegrationTestCase.php:147]: :+1:

we care to detect that there is deprecation. There is no value in having same deprecation msg once or 10 times.

File: tests/TestCase.php [hunk 3]
```diff
@@ -35,9 +35,7 @@ abstract class TestCase extends BaseTestCase
     protected function tearDown(): void
     {
         if (null !== $this->previouslyDefinedErrorHandler) {
-            foreach ($this->expectedDeprecations as $expectedDeprecation) {
-                self::assertContains($expectedDeprecation, $this->actualDeprecations);
-            }
+            self::assertSame($this->expectedDeprecations, $this->actualDeprecations);
```
Comment (keradus) [inline tests/TestCase.php:38]: this way you depends on order of triggered deprecation. if someone done DX refactoring of method, method works exactly same way, but deprecations will get triggered in other order - this assertion will fail.

File: tests/Test/AbstractIntegrationTestCase.php (hunk 1 above)
Comment (kubawerlos) [reply tests/Test/AbstractIntegrationTestCase.php]: > Does it make sense to expect exactly the same deprecation message multiple times?

Of course not, I just wanted to show that if we would want to check the exact count (as suggested in https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7578#discussion_r1434021534) there would be multiple expectations of the same message.

File: tests/TestCase.php (hunk 3 above)
Comment (kubawerlos) [reply tests/TestCase.php:38]: Is that bad? `SymfonyTestsListener` needs `expectDeprecation` called in order they are triggered.

Comment (Wirone) [reply tests/TestCase.php:38]: Symfony's implementation aside - why order of deprecations should be interesting? In my opinion you either expect some deprecation or not, it does not matter if the executed code triggers it at the beginning or in the end.

Comment (kubawerlos) [reply tests/TestCase.php:38]: I have no problem with not caring about the order (which is how it is after the latest commit), and at the same time, I can imagine someone wanting to check the order in what deprecations were triggered (like maybe from least detailed to most detailed).

File: tests/Test/AbstractIntegrationTestCase.php (hunk 1 above)
Comment (keradus) [reply tests/Test/AbstractIntegrationTestCase.php]: that was totally _not_ my point @kubawerlos . 
my point was _what if yet another deprecation would happen_. let say, the "unexpected" one - we should fail. now we silently move on.

**Comment by keradus** [comment]
recovered https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7742

**Pull request closed without merging**